src/retrieve.js

```diff
@@ -0,0 +1,183 @@
+import MiniPromise from './ponyfills/minipromise.js';
+import base64 from './ponyfills/base64.js';
+import { assign} from './ponyfills/objects.js';
+
+const URL_PATTERN = /^([a-z]+):\/\/([a-z\-_\.]+)(\/.*)/i;
+
+function cryptography() {
+  return typeof crypto !== 'undefined' ? crypto : msCrypto;
+}
+
+/**
+ * Convert a String to an ArrayBuffer
+ *
+ * ie11 Supported
+ *
+ * @param str The String to convert to an ArrayBuffer
+ * @returns {ArrayBuffer} The resulting array buffer encoded as utf-8
+ */
+export function str2ab(str) {
+  if (typeof TextEncoder !== 'undefined') {
+    return (new TextEncoder()).encode(str).buffer;
+  }
+
+  const buf = new ArrayBuffer(str.length);
+  const bufView = new Uint8Array(buf);
+  for (let i = 0, strLen = str.length; i < strLen; i++) {
+    bufView[i] = str.charCodeAt(i);
+  }
+  return buf;
+}
+
+/**
+ * Converts an msCrypto operation to a promise if needed.
+ *
+ * @param op The op (or Promise) to convert if needed.
+ * @returns {{PromiseLike} A promise
+ */
+function cryptoOperationToPromise(op) {
+  if (op.then) {
+    return op;
+  }
+
+  return MiniPromise.createPromise(function(resolve, reject) {
+    op.oncomplete = function(evt) {
+      resolve(evt.target.result);
+    };
+
+    function rejectHandler(evt) {
+      reject(evt.toString());
+    }
+    op.onerror = rejectHandler;
+    op.onabort = rejectHandler;
+  });
+}
+
+/**
+ * Sign a String with HMAC-SHA384
+ *
+ * @param {String} key The HMAC key to use for signing
+ * @param {String} payload The String to sign
+ * @returns {PromiseLike<ArrayBuffer>} The cryptographic signature
+ */
+function sign(key, payload) {
```

Comment:
Why do we need to sign requests? Is this meant for an API key?

Reply:
Yeah, it's to support authenticated clients in v2 of the API.